### PR TITLE
School alert type exception management

### PR DESCRIPTION
**Pull request by jamesjefferies**
- [x] Allow scope to be used for content generation
- [x] Allow for simple list, creation and deletion of exceptions
- [x] Add multi-check check boxes
- [x] Add reason editing functionality
- [x] Actually nobble the creation!

### CODE REVIEW

File: app/assets/stylesheets/alert_editor.scss [hunk 1]
```diff
@@ -10,3 +10,7 @@ select.colours option[value="neutral"] {
 select.colours option[value="positive"] {
   background: $bg-positive;
 }
+
```
Comment (jamesjefferies) [inline app/assets/stylesheets/alert_editor.scss]: Bootstrap styling not really working in a list item for forms, this works, didn't want to spend more time on it!

File: app/controllers/admin/alert_types/school_alert_type_exceptions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      end
+
+      def new_multiple
+        @schools = School.all
+      end
+
+      def create_multiple
+        school_ids = params[:school_ids]
+        reasons = params[:school_reasons]
+
```
Comment (jamesjefferies) [inline app/controllers/admin/alert_types/school_alert_type_exceptions_controller.rb]: I delete them all, for the alert type and re-create them from what comes back in the form...

File: spec/system/admin/alert_type_management_ratings_spec.rb [hunk 3]
```diff
@@ -0,0 +1,235 @@
+require 'rails_helper'
```
Comment (jamesjefferies) [inline spec/system/admin/alert_type_management_ratings_spec.rb]: This is pulled out of the alert_type_management spec

File: app/controllers/admin/alert_types/school_alert_type_exceptions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      def index
+        @exceptions = SchoolAlertTypeException.where(alert_type: @alert_type)
+      end
+
+      def new_multiple
+        @schools = School.all
+      end
+
+      def create_multiple
```
Comment (jamesalmond) [inline app/controllers/admin/alert_types/school_alert_type_exceptions_controller.rb]: Just use `new` and `create`

File: app/models/alert.rb [hunk 5]
```diff
@@ -47,6 +47,8 @@ class Alert < ApplicationRecord
   enum enough_data: [:enough, :not_enough, :minimum_might_not_be_accurate], _prefix: :data
   enum relevance: [:relevant, :not_relevant, :never_relevant], _prefix: :relevance
 
+  scope :without_exception, -> { joins(:alert_type).joins('LEFT OUTER JOIN school_alert_type_exceptions ON school_alert_type_exceptions.school_id = alerts.school_id AND school_alert_type_exceptions.alert_type_id = alert_types.id').where(school_alert_type_exceptions: { school_id: nil }) }
```
Comment (jamesalmond) [inline app/models/alert.rb]: Prefer `without_exceptions`, "without exception" sounds like the opposite of what we're trying to do!

File: app/views/admin/alert_types/school_alert_type_exceptions/_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,29 @@
+<%= form_tag create_multiple_admin_alert_type_school_alert_type_exceptions_path(@alert_type), method: :put do %>
+  <% SchoolGroup.all.order(name: :asc).each do |school_group| %>
```
Comment (jamesalmond) [inline app/views/admin/alert_types/school_alert_type_exceptions/_form.html.erb]: Not keen on all the direcy class access in the view. Can we push some of it to the controller?

File: app/views/admin/alert_types/school_alert_type_exceptions/_form.html.erb [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+      </li>
+      <% school_group.schools.order(name: :asc).each do |school| %>
+        <li class='list-group-item d-flex justify-content-between align-items-center'>
+          <div class="custom-control custom-checkbox">
+            <%= check_box_tag "school_ids[]", school.id, (SchoolAlertTypeException.find_by(alert_type: @alert_type, school_id: school.id)), { id: "school_ids_#{school.id}", class: "custom-control-input multi_select_#{school_group.id}_group" } %>
+            <%= label_tag "school_ids[#{school.id}]", school.name, class: 'custom-control-label' %>
+          </div>
+          <div>
+            <%= label_tag "school_reasons_#{school.id}", 'Reason' %> :
+            <%= text_field_tag 'school_reasons[]', (SchoolAlertTypeException.find_by(alert_type: @alert_type, school_id: school.id).reason if SchoolAlertTypeException.find_by(alert_type: @alert_type, school_id: school.id)), { id: "school_reasons_#{school.id}", class: "width-400" } %>
```
Comment (jamesalmond) [inline app/views/admin/alert_types/school_alert_type_exceptions/_form.html.erb]: `school_reasons[#{shool.id}]` might give us an easier way of mapping schools to reasons in the controller?

File: app/views/admin/alert_types/school_alert_type_exceptions/_form.html.erb (hunk 6 above)
Comment (jamesjefferies) [reply app/views/admin/alert_types/school_alert_type_exceptions/_form.html.erb]: Sorted now

File: app/views/admin/alert_types/school_alert_type_exceptions/_form.html.erb (hunk 7 above)
Comment (jamesjefferies) [reply app/views/admin/alert_types/school_alert_type_exceptions/_form.html.erb]: Smashing idea, sorted now

**Pull request merged**